buildkitd/buildkitd.go

```diff
@@ -356,3 +362,44 @@ func GetAvailableImageID(ctx context.Context, image string) (string, error) {
 	}
 	return string(output), nil
 }
+
+// CheckCompatibility runs all avaliable compatibility checks before starting the buildkitd daemon.
+func CheckCompatibility(ctx context.Context, settings Settings) error {
+	isNamespaced, err := isNamespacedDocker(ctx)
+	if isNamespaced {
+		return errors.New(`user namespaces are enabled, run with --bkd-arg "--userns=host" to disable`)
+	} else if err != nil {
+		return errors.Wrap(err, "failed compatibilty check")
+	}
+
+	isRootless, err := isRootlessDocker(ctx)
+	if isRootless {
+		return errors.New(`rootless docker detected. Compatibility is limited. Run with --bkd-arg "" to give it a shot`)
+	} else if err != nil {
+		return errors.Wrap(err, "failed compatibilty check")
+	}
+
+	return nil
+}
+
+func isNamespacedDocker(ctx context.Context) (bool, error) {
+	cmd := exec.CommandContext(ctx,
+		"docker", "info", "--format={{.SecurityOptions}}")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return false, errors.Wrap(err, "get docker security info")
+	}
+
+	return strings.Contains(string(output), "name=userns"), nil
+}
+
+func isRootlessDocker(ctx context.Context) (bool, error) {
```

Comment:
Playing with it; it kinda works sometimes. Clearly needs more work but looks possible. I got it to do a simple build, but `WITH DOCKER` is a no-go. Also you will run into stuff like this when manipulating the buildkitd container:

```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:370: starting container process caused: process_linux.go:459: container init caused: write sysctl key net.ipv4.ping_group_range: write /proc/sys/net/ipv4/ping_group_range: invalid argument: unknown.
: exit status 125

```

Reply:
It's possible to use rootless overlay snapshotter in buildkit. But I think it requires some additional stuff in the buildkitd container. See eg the buildkit `rootless` build: https://github.com/moby/buildkit/blob/master/Dockerfile#L288

Reply:
Seems orthogonal to the intent here. Don't want to shave that yak in this PR. Felt reasonable to include a check for it though